packages/ai/src/chains/operations/README.md

```diff
@@ -0,0 +1,140 @@
+
+For example an aiOperation might contain style changes in the form of Tailwind CSS classes and the companion `aiOperationToWs` can turn them into Webstudio-compatible styles returining a wsOperation.
+
+Once all the wsOperations are ready they are sent to the client which has specific logic to apply each of these to the Webstudio project.
+
+##### Special Case for User Interfaces
+
+Generating quality user interfaces might be challenging with LLMs.
+
+Having to do so within a generic operations framework like the one described above makes it even more challenging.
+
+For this reason when it comes to generating user interfaces, the Operation chain instead offers an operation called [`generateTemplatePrompt`](./operations/generate-template-prompt.ts) that provides information about insertion point in the existing UI and a prompt, possibly enhanced.
+
+Another chain can use these information and a more sophisticated prompt and model to then generate the user interface and return a [`generateInsertTemplate`](./operations/generate-insert-template.ts) wsOperation. Webstudio does exactly this in a rest endpoint, using the [`template-generator`](../template-generator) chain with gpt4 instead of gpt-3.5-turbo.
+
+## Usage
+
+Server side:
+
+```typescript
+import {
+  operations,
+  templateGenerator,
+  createGptModel
+  type GPTModelMessageFormat
+} from "@webstudio-is/ai";
+
+export async function handler({ request }) {
+  const { prompt, components, jsx } = await request.json();
+
+  const model = createGptModel({
+    apiKey: process.env.OPENAI_KEY,
+    organization: process.env.OPENAI_ORG,
+    temperature: 0.2,
+    model: "gpt-3.5-turbo",
+  });
+
+  const chain = operations.createChain<GPTModelMessageFormat>();
+
+  const response = await chain({
+    model,
+    context: {
+      prompt,
+      components,
+      jsx
+    }
+  });
+
+  if (response.success === false) {
+    return response;
+  }
+
+  const promptOperations = getGenerateTemplatePromptsWsOperations(response.data);
+
+  if (promptOperations.length > 0) {
+    const model = createGptModel({
+      apiKey: process.env.OPENAI_KEY,
+      organization: process.env.OPENAI_ORG,
+      temperature: 0.2,
+      model: "gpt-4",
+    });
+    const chain = templateGenerator.createChain<GPTModelMessageFormat>();
+
+    const results = await Promise.all(promptOperations.map(operation => chain({
+      model,
+      context: {
+        prompt: operation.llmPrompt,
+        components,
+      }
+    }));
+
+    replaceGenerateTemplateWithInsertTemplateWsOperations(
+      promptOperations,
+      results,
+      response,
+    );
+  }
+
+  return {
+    success: true,
+    data: response.data,
+  };
+}
+```
+
+Client side:
+
+```tsx
+import { operations, request, type ErrorResponse } from "@webstudio-is/ai";
+import { applyOperations } from "./apply-operations";
+
+request<operations.Response>(
```

Comment:
This name gives me shivers. Removed "request" package so many times. 😄

Reply:
call it `fetchSomething` where something is specific to what we do there

Reply:
what is it specifically you are fetching?

Reply:
this sseems to be tightly coupled with openai, its not just a generic fetch for any models we might have, right? so maybe fetchOpenAI

Reply:
yeah, maybe better to split into a few clear functions